### PR TITLE
Add retry to LightningClient when handling headless apps

### PR DESCRIPTION
**Pull request by nohalon**
## What does this PR do?

We've been seeing this exception sometimes so we're adding a retry on the LightningClient instantiation when handling headless apps 

This should help apps be resilient on platform releases

Credits to @alecmerdler on debugging efforts

```2023-02-03 01:18:04,003 WARNING Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fc70cafff70>: Failed to establish a new connection: [Errno 110] Connection timed out')': /v1/projects/01g2w02nfajzas5s35epaan0a7/appinstances
INFO: Traceback (most recent call last):
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_launcher/launcher.py", line 255, in run_lightning_flow
    app._run()
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning/app/core/app.py", line 504, in _run
    done = self.run_once()
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning/app/core/app.py", line 425, in run_once
    self._update_is_headless()
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning/app/core/app.py", line 532, in _update_is_headless
    _handle_is_headless(self)
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning/app/utilities/app_helpers.py", line 582, in _handle_is_headless
    list_apps_response = client.lightningapp_instance_service_list_lightningapp_instances(project_id=project_id)
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/openapi/api/lightningapp_instance_service_api.py", line 665, in lightningapp_instance_service_list_lightningapp_instances
    ) = self.lightningapp_instance_service_list_lightningapp_instances_with_http_info(
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/openapi/api/lightningapp_instance_service_api.py", line 762, in lightningapp_instance_service_list_lightningapp_instances_with_http_info
    return self.api_client.call_api(
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/openapi/api_client.py", line 362, in call_api
    return self.__call_api(resource_path, method, path_params,
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/openapi/api_client.py", line 173, in __call_api
    response_data = self.request(method,
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/rest_client.py", line 81, in wrap
    raise err
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/rest_client.py", line 75, in wrap
    response = func(*args, **kwargs)
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/openapi/api_client.py", line 388, in request
    return self.rest_client.GET(url,
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/openapi/rest.py", line 249, in GET
    return self.request("GET",
  File "/home/zeus/.local/lib/python3.8/site-packages/lightning_cloud/openapi/rest.py", line 239, in request
    raise ApiException(http_resp=r)
lightning_cloud.openapi.rest.ApiException: (502)
Reason: Bad Gateway
HTTP response headers: HTTPHeaderDict({'Content-Type': 'text/html; charset=UTF-8', 'Referrer-Policy': 'no-referrer', 'Content-Length': '332', 'Date': 'Fri, 03 Feb 2023 11:57:57 GMT', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000'})
HTTP response body: b'\n<html><head>\n<meta http-equiv="content-type" content="text/html;charset=utf-8">\n<title>502 Server Error</title>\n</head>\n<body text=#000000 bgcolor=#ffffff>\n<h1>Error: Server Error</h1>\n<h2>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.</h2>\n<h2></h2>\n</body></html>\n'
```

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

cc @borda

### CODE REVIEW

**Review by alecmerdler — APPROVED** [review]
Nice!

File: src/lightning/app/utilities/app_helpers.py [hunk 1]
```diff
@@ -592,7 +592,7 @@ def _handle_is_headless(app: "LightningApp"):
 
     from lightning.app.utilities.network import LightningClient
 
-    client = LightningClient()
+    client = LightningClient(retry=True)
```
Comment (hhsecond) [inline src/lightning/app/utilities/app_helpers.py:595]: This doesn't help to solve the problem though. The default value for `retry` is already `True`

**Review by hhsecond — APPROVED** [review]

File: src/lightning/app/utilities/app_helpers.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/app/utilities/app_helpers.py:595]: Yes, this isn't the fix. True is the default value.

**Comment by awaelchli** [comment]
#16382 didn't get released yet. So this PR is not necessary on top of it and the next version release will address this.

**Comment by awaelchli** [comment]
The change that flipped the default to True was released last week #16382. Since there hasn't been any activity, I'm closing the PR. If it was a mistake, let me know.

**Pull request closed without merging**